components/fxa-client/src/lib.rs

```diff
@@ -2,597 +2,1268 @@
+//!
+//! * When the user wants to sign in to your application, direct them through
+//!   a web-based OAuth flow using [`begin_oauth_flow`](FirefoxAccount::begin_oauth_flow)
+//!   or [`begin_pairing_flow`](FirefoxAccount::begin_pairing_flow); when they return
+//!   to your registered `redirect_uri`, pass the resulting authorization state back to
+//!   [`complete_oauth_flow`](FirefoxAccount::complete_oauth_flow) to sign them in.
+//!
+//! * Display information about the signed-in user by using the data from
+//!   [`get_profile`](FirefoxAccount::get_profile).
+//!
+//! * Access account-related services on behalf of the user by obtaining OAuth
+//!   access tokens via [`get_access_token`](FirefoxAccount::get_access_token).
+//!
+//! * If the user opts to sign out of the application, calling [`disconnect`](FirefoxAccount::disconnect)
+//!   and then discarding any persisted account data.
+
 #![allow(unknown_lints)]
 #![warn(rust_2018_idioms)]
-
-use crate::{
-    commands::send_tab::SendTabPayload,
-    device::Device,
-    oauth::{AuthCircuitBreaker, OAuthFlow, OAUTH_WEBCHANNEL_REDIRECT},
-    scoped_keys::ScopedKey,
-    state_persistence::State,
-};
-pub use crate::{
-    config::Config,
-    error::*,
-    oauth::{AccessTokenInfo, IntrospectInfo, RefreshToken},
-    profile::Profile,
-    telemetry::FxaTelemetry,
-};
 use serde_derive::*;
-use std::{
-    cell::RefCell,
-    collections::{HashMap, HashSet},
-    sync::Arc,
-};
-use url::Url;
-
-#[cfg(feature = "integration_test")]
-pub mod auth;
-mod commands;
-mod config;
-pub mod device;
-pub mod error;
-pub mod ffi;
-mod http_client;
-pub mod migrator;
-// Include the `msg_types` module, which is generated from msg_types.proto.
-pub mod msg_types {
-    include!("mozilla.appservices.fxaclient.protobuf.rs");
+use std::collections::HashMap;
+use std::convert::{TryFrom, TryInto};
+use thiserror::Error;
+
+// All the implementation details live in this "internal" module.
+// Aspirationally, I'd like to make it private, so that the public API of the crate
+// is entirely the same as the API exposed to consumers via UniFFI. That's not
+// possible right now because some of our tests/example use features that we do
+// not currently expose to consumers. But we should figure out how to expose them!
+pub mod internal;
+
+uniffi_macros::include_scaffolding!("fxa_client");
+
+/// Generic error type thrown by many [`FirefoxAccount`] operations.
+///
+/// Precise details of the error are hidden from consumers, mostly due to limitations of
+/// how we expose this API to other languages. The type of the error indicates how the
+/// calling code should respond.
+///
+#[derive(Debug, Error)]
+pub enum FxaError {
+    /// Thrown when there was a problem with the authentication status of the account,
+    /// such as an expired token. The application should [check its authorization status](
+    /// FirefoxAccount::check_authorization_status) to see whether it has been disconnected,
```

Comment:
If you use a new-enough `cargo`, then `cargo doc` will automatically linkify references to structs and methods and what-not in this form. It's pretty nice!